src/Env.hs

```diff
@@ -0,0 +1,320 @@
+                             ylsop  :: [Y],
+                             klsop  :: [K],
+                             tlsop  :: [T],
+                             top  :: T}
+                 deriving (Show, Eq, Typeable)
+
+data PredicateEnv = Pred1 Predicate1
+                  | Pred2 Predicate2
+                  deriving (Show, Eq, Typeable)
+
+data Predicate1 = Predicate1 { namepred1 :: String,
+                               ylspred1  :: [Y],
+                               klspred1  :: [K],
+                               tlspred1  :: [T],
+                               ppred1    :: Prop}
+                  deriving (Show, Eq, Typeable)
+
+data Predicate2 = Predicate2 { namepred2 :: String,
+                               plspred2  :: [Prop],
+                               ppred2    :: Prop }
+                  deriving (Show, Eq, Typeable)
+
+data VarEnv = VarEnv { typeConstructors :: M.Map String TypeConstructor,
+                       varConstructors  :: M.Map String VarConstructor,
+                       operations       :: M.Map String Operation,
+                       predicates       :: M.Map String PredicateEnv,
+                       typeVarMap       :: M.Map TypeVar Type,
+                       varMap           :: M.Map Var T,
+                       names            :: [String],  -- a global list which contains all the names of types in that env
+                       declaredNames    :: [String],  -- a global list which contains all the names of elements declared in that env
+                       errors           :: String }   -- a string which accumulates all the errors founded during the run of the typechecker
+              deriving (Show, Eq, Typeable)
+
+
+checkTypeVar :: VarEnv -> TypeVar -> VarEnv
+checkTypeVar e v = if (M.member v (typeVarMap e))
+                   then e
+                   else e { errors = (errors e) ++ ("TypeVar " ++ (show v) ++ "is not in scope \n") }
+
+checkVar :: VarEnv -> Var -> VarEnv
+checkVar e v = if (M.member v (varMap e))
+               then e
+               else e { errors = (errors e) ++ ("Var " ++ (show v) ++ "is not in scope \n") }
+
+
+checkY :: VarEnv -> Y -> VarEnv
+checkY e (TypeVarY y) = checkTypeVar e y
+checkY e (VarY y)     = checkVar e y
+
+checkArg :: VarEnv -> Arg -> VarEnv
+checkArg e (AVar v) = checkVar e v
+checkArg e (AT (TConstr i)) = if (nameCons i) `elem` (declaredNames e)
+                              then checkVar e (VarConst (nameCons i))
+                              else checkT e (TConstr i)
+checkArg e (AT t) = checkT e t
+
+checkT :: VarEnv -> T -> VarEnv
+checkT e (TTypeVar t) = checkTypeVar e t
+checkT e (TConstr c)  = checkConstructorInvoker e c
+
+checkType :: VarEnv -> Type -> VarEnv
+checkType e t = e
+
+-- --TODO: fix back...
+-- checkConstructorInvoker :: VarEnv -> ConstructorInvoker -> VarEnv
+-- checkConstructorInvoker e const = e
+
+checkConstructorInvoker :: VarEnv -> ConstructorInvoker -> VarEnv
```

Comment:
this function's name would also need to be changed if we change ConstructorInvoker